File: core/src/main/java/javabot/dao/FactoidDao.java
package javabot.dao;

import com.mongodb.WriteResult;
import javabot.dao.util.QueryParam;
import javabot.model.Factoid;
import javabot.model.Persistent;
import javabot.model.criteria.FactoidCriteria;
import org.mongodb.morphia.query.Query;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import java.time.LocalDateTime;
import java.util.List;

@SuppressWarnings({"ConstantNamingConvention"})
public class FactoidDao extends BaseDao<Factoid> {
    private static final Logger log = LoggerFactory.getLogger(FactoidDao.class);

    @Inject
    private ChangeDao changeDao;

    @Inject
    private ConfigDao dao;

    public FactoidDao() {
        super(Factoid.class);
    }

    public void save(final Persistent persistent) {
        Factoid factoid = (Factoid) persistent;
        Factoid old = find(persistent.getId());
        super.save(persistent);
        if (old != null) {
            changeDao.logChange(String.format("%s changed '%s' from '%s' to '%s'",
                                              factoid.getUserName(), factoid.getName(), old.getValue(), factoid.getValue()));
        } else {
            changeDao.logChange(String.format("%s added '%s' with '%s'", factoid.getUserName(), factoid.getName(),
                                              factoid.getValue()));
        }
    }

    public boolean hasFactoid(final String key) {
        FactoidCriteria criteria = new FactoidCriteria(ds);
        criteria.upperName().equal(key.toUpperCase());
        return criteria.query().get() != null;
    }

    public Factoid addFactoid(final String sender, final String key, final String value) {
        final Factoid factoid = new Factoid();
        factoid.setId(factoid.getId());
        factoid.setName(key);
        factoid.setValue(value);
        factoid.setUserName(sender);
        factoid.setUpdated(LocalDateTime.now());
        factoid.setLastUsed(LocalDateTime.now());
        save(factoid);
        changeDao.logAdd(sender, key, value);
        return factoid;
    }

    public void delete(final String sender, final String key) {
        final Factoid factoid = getFactoid(key);
        if (factoid != null) {
            delete(factoid.getId());
            changeDao.logChange(String.format("%s removed '%s' with a value of '%s'", sender, key, factoid.getValue()));
        }
    }

    public Factoid getFactoid(final String name) {
        FactoidCriteria criteria = new FactoidCriteria(ds);
        criteria.upperName().equal(name.toUpperCase());
        Factoid factoid = criteria.query().get();
        if (factoid != null) {
            factoid.setLastUsed(LocalDateTime.now());
            super.save(factoid);
        }
        return factoid;
    }

    public Factoid getParameterizedFactoid(final String name) {
        FactoidCriteria criteria = new FactoidCriteria(ds);
        criteria.or(
                       criteria.upperName().equal(name.toUpperCase() + " $1"),
                       criteria.upperName().equal(name.toUpperCase() + " $^"),
                       criteria.upperName().equal(name.toUpperCase() + " $+")
                   );
        Factoid factoid = criteria.query().get();
        if (factoid != null) {
            factoid.setLastUsed(LocalDateTime.now());
            super.save(factoid);
        }
        return factoid;
    }

    public Long count() {
        return ds.createQuery(Factoid.class).countAll();
    }

    public Long countFiltered(final Factoid filter) {
        return buildFindQuery(null, filter, true).countAll();
    }

    public List<Factoid> getFactoidsFiltered(final QueryParam qp, final Factoid filter) {
        return buildFindQuery(qp, filter, false).asList();
    }

    private Query<Factoid> buildFindQuery(final QueryParam qp, final Factoid filter, final boolean count) {
        FactoidCriteria criteria = new FactoidCriteria(ds);
        if (filter.getName() != null) {
            criteria.upperName().contains(filter.getName().toUpperCase());
        }
        if (filter.getUserName() != null) {
            criteria.upperUserName().contains(filter.getUserName().toUpperCase());
        }
        if (filter.getValue() != null) {
            criteria.upperValue().contains(filter.getValue().toUpperCase());
        }
        if (!count && qp != null && qp.hasSort()) {
            criteria.query().order((qp.isSortAsc() ? "" : "-") + "upper" + qp.getSort());
            criteria.query().offset(qp.getFirst());
            criteria.query().limit(qp.getCount());
        }
        return criteria.query();
    }

    public WriteResult deleteAll() {
        return ds.delete(ds.createQuery(Factoid.class));
    }
}

File: src/main/kotlin/javabot/dao/FactoidDao.kt
package javabot.dao

import com.antwerkz.sofia.Sofia
import com.mongodb.client.result.DeleteResult
import dev.morphia.Datastore
import dev.morphia.DeleteOptions
import dev.morphia.query.FindOptions
import dev.morphia.query.Query
import dev.morphia.query.filters.Filters.eq
import dev.morphia.query.filters.Filters.or
import java.time.LocalDateTime
import java.util.Locale
import java.util.regex.PatternSyntaxException
import javabot.dao.util.QueryParam
import javabot.model.Factoid
import javabot.model.Persistent
import javabot.model.criteria.FactoidCriteria.Companion.upperName
import javabot.model.criteria.FactoidCriteria.Companion.upperUserName
import javax.inject.Inject

class FactoidDao
@Inject
constructor(ds: Datastore, var changeDao: ChangeDao, var configDao: ConfigDao) :
    BaseDao<Factoid>(ds, Factoid::class.java) {

    override fun save(entity: Persistent) {
        val factoid = entity as Factoid
        if (factoid.name == "") {
            throw IllegalArgumentException(Sofia.factoidCantBeBlank("name"))
        }
        if (factoid.value == "") {
            throw IllegalArgumentException(Sofia.factoidCantBeBlank("value"))
        }
        if (factoid.userName == "") {
            throw IllegalArgumentException(Sofia.factoidCantBeBlank("user name"))
        }
        super.save(entity)
    }

    fun hasFactoid(key: String) =
        ds.find(Factoid::class.java)
            .filter(upperName().eq(key.uppercase(Locale.getDefault())))
            .first() != null

    fun addFactoid(sender: String, key: String, value: String, location: String): Factoid {
        return addFactoid(sender, key, value, location, LocalDateTime.now())
    }

    fun addFactoid(
        sender: String,
        key: String,
        value: String,
        location: String,
        updated: LocalDateTime,
    ): Factoid {
        val factoid = Factoid(key, value, sender)
        factoid.updated = updated
        factoid.lastUsed = LocalDateTime.now()
        save(factoid)
        changeDao.logFactoidAdded(sender, key, value, location)
        return factoid
    }

    fun delete(sender: String, key: String, location: String) {
        val factoid = getFactoid(key)
        if (factoid != null) {
            delete(factoid.id)
            changeDao.logFactoidRemoved(sender, key, factoid.value, location)
        }
    }

    fun getFactoid(name: String): Factoid? {
        val criteria =
            ds.find(Factoid::class.java).filter(upperName().eq(name.uppercase(Locale.getDefault())))
        val factoid = criteria.first()
        if (factoid != null) {
            factoid.lastUsed = LocalDateTime.now()
            super.save(factoid)
        }
        return factoid
    }

    fun getParameterizedFactoid(name: String): Factoid? {
        val factoid =
            ds.find(Factoid::class.java)
                .filter(
                    or(
                        upperName().eq(name.uppercase(Locale.getDefault()) + " \$1"),
                        upperName().eq(name.uppercase(Locale.getDefault()) + " \$^"),
                        upperName().eq(name.uppercase(Locale.getDefault()) + " \$+"),
                    )
                )
                .first()
        if (factoid != null) {
            factoid.lastUsed = LocalDateTime.now()
            super.save(factoid)
        }
        return factoid
    }

    fun count(): Long {
        return ds.find(Factoid::class.java).count()
    }

    fun countFiltered(filter: Factoid): Long {
        return buildFindQuery(filter).count()
    }

    fun getFactoidsFiltered(qp: QueryParam, filter: Factoid): List<Factoid> {
        val query = buildFindQuery(filter)
        val options = FindOptions().skip(qp.first).limit(qp.count)
        if (qp.hasSort()) {
            options.sort(qp.toSort("upper"))
        }
        return query.iterator(options).toList()
    }

    private fun buildFindQuery(filter: Factoid): Query<Factoid> {
        val query = ds.find(Factoid::class.java)
        if (filter.name != "") {
            try {
                query.filter(upperName().eq(filter.name.uppercase(Locale.getDefault())))
            } catch (e: PatternSyntaxException) {
                Sofia.logFactoidInvalidSearchValue(filter.name)
            }
        }

        if (filter.userName != "") {
            try {
                query.filter(upperUserName().eq(filter.userName.uppercase(Locale.getDefault())))
            } catch (e: PatternSyntaxException) {
                Sofia.logFactoidInvalidSearchValue(filter.userName)
            }
        }

        if (filter.value != "") {
            try {
                query.filter(eq("upperValue", filter.value.uppercase(Locale.getDefault())))
            } catch (e: PatternSyntaxException) {
                Sofia.logFactoidInvalidSearchValue(filter.value)
            }
        }

        return query
    }

    fun deleteAll(): DeleteResult {
        return ds.find(Factoid::class.java).delete(DeleteOptions().multi(true))
    }
}
